toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexService.kt

```diff
@@ -0,0 +1,233 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.crates.local
+
+import com.google.gson.Gson
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.components.PersistentStateComponent
+import com.intellij.openapi.components.State
+import com.intellij.openapi.components.Storage
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.Task
+import com.intellij.util.io.*
+import org.eclipse.jgit.lib.ObjectId
+import org.eclipse.jgit.lib.Repository
+import org.eclipse.jgit.revwalk.RevTree
+import org.eclipse.jgit.revwalk.RevWalk
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder
+import org.eclipse.jgit.treewalk.TreeWalk
+import org.rust.openapiext.pluginDirInSystem
+import org.rust.stdext.cleanDirectory
+import java.io.BufferedReader
+import java.io.DataInput
+import java.io.DataOutput
+import java.io.InputStreamReader
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.concurrent.atomic.AtomicReference
+
+@State(name = "CratesLocalIndexState", storages = [Storage("rust.crateslocalindex.xml")])
+class CratesLocalIndexService : PersistentStateComponent<CratesLocalIndexService.Companion.CratesLocalIndexState>, Disposable {
+    private val baseCratesLocalRegistryDir: Path
+        get() = pluginDirInSystem().resolve("crates-local-index")
+
+    private val corruptionMarkerFile: Path
+        get() = baseCratesLocalRegistryDir.resolve(CORRUPTION_MARKER_NAME)
+
+    private val userCargoIndexDir: Path
+        get() = Paths.get(System.getProperty("user.home"), CARGO_REGISTRY_INDEX_LOCATION)
+
+    private val repository: Repository = FileRepositoryBuilder()
+        .setGitDir(userCargoIndexDir.toFile())
+        .build()
```

Comment:
Probably we should handle `RepositoryNotFoundException` somehow

Reply:
Maybe in the future here we could throw an exception to handle it inside, for instance, crate resolver

Reply:
OK, let's leave something like `// TODO: handle RepositoryNotFoundException` for now